client/relay-chain-rpc-interface/src/reconnecting_ws_client.rs

```diff
@@ -0,0 +1,495 @@
+		let import_subscription = self
+			.active_client
+			.subscribe::<RelayHeader, _>(
+				"chain_subscribeAllHeads",
+				rpc_params![],
+				"chain_unsubscribeAllHeads",
+			)
+			.await
+			.map_err(|e| {
+				tracing::error!(target: LOG_TARGET, ?e, "Unable to open subscription.");
+				e
+			})?;
+		let best_subscription = self
+			.active_client
+			.subscribe::<RelayHeader, _>(
+				"chain_subscribeNewHeads",
+				rpc_params![],
+				"chain_unsubscribeNewHeads",
+			)
+			.await
+			.map_err(|e| {
+				tracing::error!(target: LOG_TARGET, ?e, "Unable to open subscription.");
+				e
+			})?;
+		let finalized_subscription = self
+			.active_client
+			.subscribe::<RelayHeader, _>(
+				"chain_subscribeFinalizedHeads",
+				rpc_params![],
+				"chain_unsubscribeFinalizedHeads",
+			)
+			.await
+			.map_err(|e| {
+				tracing::error!(target: LOG_TARGET, ?e, "Unable to open subscription.");
+				e
+			})?;
+
+		Ok(RelayChainSubscriptions {
+			import_subscription,
+			best_subscription,
+			finalized_subscription,
+		})
+	}
+
+	/// Create a request future that performs an RPC request and sends the results to the caller.
+	/// In case of a dead websocket connection, it returns the original request parameters to
+	/// enable retries.
+	fn create_request(
+		&self,
+		method: String,
+		params: ArrayParams,
+		response_sender: OneshotSender<Result<JsonValue, JsonRpseeError>>,
+	) -> impl Future<Output = Result<(), RpcDispatcherMessage>> {
+		let future_client = self.active_client.clone();
+		async move {
+			let resp = future_client.request(&method, params.clone()).await;
+
+			// We should only return the original request in case
+			// the websocket connection is dead and requires a restart.
+			// Other errors should be forwarded to the request caller.
+			if let Err(JsonRpseeError::RestartNeeded(_)) = resp {
+				return Err(RpcDispatcherMessage::Request(method, params, response_sender))
+			}
+
+			if let Err(err) = response_sender.send(resp) {
+				tracing::debug!(
+					target: LOG_TARGET,
+					?err,
+					"Recipient no longer interested in request result"
+				);
+			}
+			Ok(())
+		}
+	}
+}
+
+impl ReconnectingWebsocketWorker {
+	/// Create new worker. Returns the worker and a channel to register new listeners.
+	async fn new(
+		urls: Vec<Url>,
+	) -> (ReconnectingWebsocketWorker, TokioSender<RpcDispatcherMessage>) {
+		let (tx, rx) = tokio_channel(100);
+		let worker = ReconnectingWebsocketWorker {
+			ws_urls: urls,
+			client_receiver: rx,
+			self_sender: tx.clone(),
+			imported_header_listeners: Vec::new(),
+			finalized_header_listeners: Vec::new(),
+			best_header_listeners: Vec::new(),
+		};
+		(worker, tx)
+	}
+
+	///	Reconnect via [`ClientManager`] and provide new notification streams.
+	async fn handle_reconnect(
+		&mut self,
+		client_manager: &mut ClientManager,
+		pending_requests: &mut FuturesUnordered<
+			impl Future<Output = Result<(), RpcDispatcherMessage>>,
+		>,
```

Comment:
```suggestion
		>,
		first_failed: Request,
```

The request that we stored in `should_reconnect`.